src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -1702,72 +1710,84 @@ private void handleSwitch(JCTree switchTree,
+                                        }
+                                    } else if (!stringSwitch && !types.isAssignable(seltype, syms.intType)) {
+                                        log.error(pat.pos(), Errors.ConstantLabelNotCompatible(pattype, seltype));
+                                    } else if (!labels.add(pattype.constValue())) {
+                                        log.error(c.pos(), Errors.DuplicateCaseLabel);
                                     } else {
-                                        log.error(expr.pos(),
-                                                  (stringSwitch ? Errors.StringConstReq : Errors.ConstExprReq));
+                                        checkCaseLabelDominated(pat.pos(), coveredTypes, types.boxedTypeOrType(pattype));
                                     }
-                                } else if (!stringSwitch && !types.isAssignable(seltype, syms.intType)) {
-                                    log.error(pat.pos(), Errors.ConstantLabelNotCompatible(pattype, seltype));
-                                } else if (!labels.add(pattype.constValue())) {
-                                    log.error(c.pos(), Errors.DuplicateCaseLabel);
-                                } else {
-                                    checkCaseLabelDominated(pat.pos(), coveredTypes, types.boxedTypeOrType(pattype));
                                 }
                             }
+                            if (wasPattern && !wasNonEmptyFallThrough) {
+                                log.error(pat.pos(), Errors.FlowsThroughFromPattern);
+                            }
+                            wasConstant = true;
                         }
                     } else if (pat.hasTag(DEFAULTCASELABEL)) {
                         if (hasDefault) {
                             log.error(pat.pos(), Errors.DuplicateDefaultLabel);
                         } else if (hasTotalPattern) {
                             log.error(pat.pos(), Errors.TotalPatternAndDefault);
-                        } else if (matchBindings.bindingsWhenTrue.nonEmpty()) {
-                            //there was a pattern, and the execution flows into a default:
+                        } else if (wasPattern && !wasNonEmptyFallThrough) {
                             log.error(pat.pos(), Errors.FlowsThroughFromPattern);
                         }
                         hasDefault = true;
+                        wasDefault = true;
                         matchBindings = MatchBindingsComputer.EMPTY;
                     } else {
-                        if (prevCompletedNormally) {
+                        boolean isTypePattern = pat.hasTag(BINDINGPATTERN);
+                        if (wasPattern || wasConstant || wasDefault ||
+                            (wasNullPattern && (!isTypePattern || wasNonEmptyFallThrough))) {
                             log.error(pat.pos(), Errors.FlowsThroughToPattern);
```

Comment:
Is this a good error message for the condition? E.g. it seems like here we're mostly complaining about some illegal mix of constant, default and pattern labels - which seems more something to do with well-formedness than with fallthrough?